jxl/src/render/simple_pipeline/save.rs

```diff
@@ -3,374 +3,282 @@
-        let buf_size = if orientation.is_transposing() {
-            (size.1, size.0)
-        } else {
-            size
-        };
-        Ok(SaveStage {
-            stage_type,
-            channel,
-            buf: Mutex::new(Image::new(buf_size)?),
-            scale,
-            orientation,
-        })
-    }
-
-    pub(crate) fn new_with_buffer(
-        stage_type: SaveStageType,
-        channel: usize,
-        img: Image<T>,
-        scale: T,
+        channels: &[usize],
         orientation: Orientation,
-    ) -> SaveStage<T> {
-        SaveStage {
-            stage_type,
-            channel,
-            buf: Mutex::new(img),
-            scale,
+        output_buffer_index: usize,
+        mut color_type: JxlColorType,
+        data_format: JxlDataFormat,
+    ) -> SaveStage {
+        let mut channels = channels.to_vec();
+        if color_type == JxlColorType::Bgr {
+            color_type = JxlColorType::Rgb;
+            channels.swap(0, 2);
+        }
+        if color_type == JxlColorType::Bgra {
+            color_type = JxlColorType::Rgba;
+            channels.swap(0, 2);
+        }
+        Self {
+            channels,
             orientation,
+            output_buffer_index,
+            color_type,
+            data_format,
         }
     }
-
-    pub(crate) fn buffer(&self) -> impl std::ops::Deref<Target = Image<T>> {
-        self.buf.lock().unwrap()
-    }
-
-    pub(crate) fn into_buffer(self) -> Image<T> {
-        self.buf.into_inner().unwrap()
-    }
 }
 
-impl<T: ImageDataType> std::fmt::Display for SaveStage<T> {
+impl std::fmt::Display for SaveStage {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         write!(
             f,
-            "save channel {} (type {:?}) scale {:?}",
-            self.channel,
-            T::DATA_TYPE_ID,
-            self.scale,
+            "save channels {:?} (type {:?} {:?})",
+            self.channels, self.color_type, self.data_format
         )
     }
 }
 
-impl<T: ImageDataType> SaveStage<T> {
-    fn process_row_chunk(
+impl SaveStage {
+    pub(super) fn save(
         &self,
-        position: (usize, usize),
-        mut xsize: usize,
-        row: &mut [&[T]],
-        _state: Option<&mut dyn std::any::Any>,
-    ) {
-        let input = row[0];
-        let mut buf = self.buf.lock().unwrap();
-        let mut outbuf_rect = buf.as_rect_mut();
-        let (out_w, out_h) = outbuf_rect.size();
-
-        // Establish source dimensions based on the orientation.
-        let (w_src, h_src) = if self.orientation.is_transposing() {
-            (out_h, out_w) // Swapped
+        data: &[Image<f64>],
+        buffers: &mut [Option<JxlOutputBuffer>],
+    ) -> Result<()> {
+        for i in self.channels.iter().skip(1) {
+            assert_eq!(data[self.channels[0]].size(), data[*i].size());
+        }
+        let Some(buf) = buffers[self.output_buffer_index].as_mut() else {
+            return Ok(());
```

Comment:
why is it not an error?

Reply:
Because the user might decide to not provide buffers for some calls (i.e. because they are trying to get jxl-info-like functionality), and we already check elsewhere that they provide the correct number of buffers if they do provide them.